base/compiler/abstractinterpretation.jl

```diff
@@ -860,27 +862,43 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
     end
     # From here on, we start asking for results on the widened types, rather than
     # the precise (potentially const) state type
-    statetype = widenconst(statetype)
-    valtype = widenconst(valtype)
+    # statetype and valtype are reinitialized in the first iteration below from the
+    # (widened) stateordonet, which has not yet been fully analyzed in the loop above
+    statetype = Bottom
+    valtype = Bottom
+    may_have_terminated = Nothing <: stateordonet
     while valtype !== Any
-        stateordonet = abstract_call_known(interp, iteratef, nothing, Any[Const(iteratef), itertype, statetype], sv).rt
-        stateordonet = widenconst(stateordonet)
-        nounion = typesubtract(stateordonet, Nothing, 0)
-        if !isa(nounion, DataType) || !(nounion <: Tuple) || isvatuple(nounion) || length(nounion.parameters) != 2
+        nounion = typeintersect(typesubtract(stateordonet, Nothing, 0), Tuple{Any,Any})
+        if nounion === Union{}
+            # handle iterator failing or giving an invalid answer below
+            nounion = Tuple{Union{}, Union{}}
+        elseif !isa(nounion, DataType) || !(nounion <: Tuple) || isvatuple(nounion) || length(nounion.parameters) != 2
+            # nounion is of a type we cannot handle
             valtype = Any
             break
         end
         if nounion.parameters[1] <: valtype && nounion.parameters[2] <: statetype
+            # reached a fixpoint or iterator failed/gave invalid answer
             if typeintersect(stateordonet, Nothing) === Union{}
```

Comment:
```suggestion
            if !(Nothing <: stateordonet)
```
As `Nothing` is a leaf type, this should be equal but more efficient?

Reply:
It should be essential identical (since typeintersect can check that property too), so not worth worrying about the difference too much